### ISSUE TITLE
Allow to customize location of .eggs directory

### ISSUE THREAD

**Issue by tgpfeiffer** [pypa/setuptools#1854]
Apparently the PYTHON_EGG_CACHE environment variable does not affect the location of the `.eggs` directory when building a Python package using a command like `python setup.py build`: https://github.com/pypa/setuptools/blob/38b9010bc3527e89561c7a985e0908a3071b7ff5/setuptools/dist.py#L745

My code lives in a read-only partition mounted into a Docker container and therefore the build fails:
```
...
  File "/usr/local/lib/python3.6/dist-packages/pkg_resources/__init__.py", line 1077, in obtain
    return installer(requirement)
  File "/usr/local/lib/python3.6/dist-packages/setuptools/dist.py", line 778, in fetch_build_egg
    install_dir = self.get_egg_cache_dir()
  File "/usr/local/lib/python3.6/dist-packages/setuptools/dist.py", line 747, in get_egg_cache_dir
    os.mkdir(egg_cache_dir)
OSError: [Errno 30] Read-only file system: './.eggs'
```
Setting the PYTHON_EGG_CACHE variable has no effect on this behavior, but I guess it should?

**Comment by gilbsgilbs** [comment]
@tgpfeiffer Did you find a workaround? I would also benefit from such feature for slightly different reasons. 

I have two docker containers that mount the same directory for local development purposes. They both do a `pip install -e` at the same time, but one them them fails because of a race condition; they try to create the setuptools egg at the same time which leads to the following error for one of them:

```
ERROR: Command errored out with exit status 1:
 command: /venv/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/app/setup.py'"'"'; __file__='"'"'/app/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info
     cwd: /app/
Complete output (27 lines):
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/app/setup.py", line 74, in <module>
    'Werkzeug[watchdog]',
  File "/venv/lib/python3.5/site-packages/setuptools/__init__.py", line 144, in setup
    _install_setup_requires(attrs)
  File "/venv/lib/python3.5/site-packages/setuptools/__init__.py", line 139, in _install_setup_requires
    dist.fetch_build_eggs(dist.setup_requires)
  File "/venv/lib/python3.5/site-packages/setuptools/dist.py", line 721, in fetch_build_eggs
    replace_conflicting=True,
  File "/venv/lib/python3.5/site-packages/pkg_resources/__init__.py", line 782, in resolve
    replace_conflicting=replace_conflicting
  File "/venv/lib/python3.5/site-packages/pkg_resources/__init__.py", line 1065, in best_match
    return self.obtain(req, installer)
  File "/venv/lib/python3.5/site-packages/pkg_resources/__init__.py", line 1077, in obtain
    return installer(requirement)
  File "/venv/lib/python3.5/site-packages/setuptools/dist.py", line 777, in fetch_build_egg
    return fetch_build_egg(self, req)
  File "/venv/lib/python3.5/site-packages/setuptools/installer.py", line 133, in fetch_build_egg
    wheel.install_as_egg(dist_location)
  File "/venv/lib/python3.5/site-packages/setuptools/wheel.py", line 102, in install_as_egg
    self._install_as_egg(destination_eggdir, zf)
  File "/venv/lib/python3.5/site-packages/setuptools/wheel.py", line 110, in _install_as_egg
    self._convert_metadata(zf, destination_eggdir, dist_info, egg_info)
  File "/venv/lib/python3.5/site-packages/setuptools/wheel.py", line 131, in _convert_metadata
    os.mkdir(destination_eggdir)
FileExistsError: [Errno 17] File exists: '/app/.eggs/setuptools_scm-3.3.3-py3.5.egg'
```

Because setuptools doesn't use `PYTHON_EGG_CACHE` in this case, I can't put the `.eggs` directory inside my containers which is what I'd really want to do. My current workaround is to create two directories `.eggs_container0` and `.eggs_container1` locally and mount them in the containers using volumes (`docker run -v .eggs_container0:/app/.eggs`). That's kind of clumsy and it messes up my local directories for no reason.

@jaraco What do you think is the proper fix? Would it be acceptable to create the `.eggs` directory in `PYTHON_EGG_CACHE` directory if the environment variable is set? If so, I could maybe submit a PR.

**Comment by benoit-pierre** [comment]
There's no relation between `PYTHON_EGG_CACHE` (the default cache location for extracting resources from zipped eggs) and the `.eggs` directory used by setuptools' `fetch_build_egg` for installing missing `setup_requires`/`tests_require` requirements.

setuptools is not multi-process / thread safe, you can't expect to be able to run concurrent commands without issues, which means that using editable installs is similarly not multi-process / thread safe, and you'll have to synchronize concurrent installs yourself.

Note that the issue is not just about .eggs, you could run into issues with concurrent generations of the egg-info metadata.

**Comment by gilbsgilbs** [comment]
That makes sense. I'll try to find a synchronization mechanism then. Thanks.

**Comment by gilbsgilbs** [comment]
@benoit-pierre Sorry for taking your time on this, however, by closing this issue, are you considering that installing `.eggs` directory somewhere else than the project directory is not a use-case setuptools maintainers want to address? Unless I'm missing something obvious, I don't think installing a python package in editable mode while its source being read-only is particularly illegitimate.

More specifically, if my current working directory is NOT where the `setup.py` lays, `pip install -e /absolute/path/to/my/project/` would still create try to create the `.eggs` directory in the project's directory (and not in the current working directory). Isn't this a bug (maybe not)?

**Comment by benoit-pierre** [comment]
It's not a valid use case, because you need write support to generate the egg-info metadata. And yes, `.eggs` and `.egg-info` directory are generated in the source directory.

**Comment by gilbsgilbs** [comment]
> It's not a valid use case, because you need write support to generate the egg-info metadata. And yes, .eggs and .egg-info directory are generated in the source directory.

The point of this issue is that you could write them both in the current working directory (or somewhere else) instead of the project's directory if write isn't available there. Is there a specific reason why this would not be possible or why it would not be disable to implement it? Apart from saying "it's not a valid use-case" while not suggesting a portable solution to a use-case that does exists for at least two people (and I suspect a lot more).

**Comment by craigphicks** [comment]
@gilbsgilbs  @tgpfeiffer 
I had a slightly different problem when specifying an install directory 
```
$ easy_install --install-dir /opt/myapp  myapp.egg
<fail message>
```
it erred out with a message that it could only write to a valid `site-packages` directory.
As I workaround I set up a virtual environment and activated it
```
$ python3 -m venv --prompt myapp /opt/myapp
$ /source opt/myapp/bin/activate
(myapp)$ easy_install --install-dir /opt/myapp  myapp.egg
<success message>
```
As a result under `/opt/myapp/lib/site-packages/lib` were mostly many compact '.egg' files, and under '`/opt/myapp/share` were many compact wheel files.  The total memory used by the virtual environment only (containing python3, etc.) was 8mb, a small fraction of the total compact installed 'myapp' which came to 84mb.

Of course to call 'myapp' pre-requires another call to
'''  
$ /source opt/myapp/bin/activate
'''
I don't grasp your exact usage requirements, but there seemed enough potential overlap I thought it might be worth posting this message.

**Comment by wagenadl** [comment]
I'd like to see this re-opened. I use separation between source and build directories for all my projects. "python -m build" with setuptools as a backend supports this workflow perfectly, except that the automatically generated .egg-info folder is created in my source rather than in my build directory. Since the egg-info is automatically generated, there is no reason for it to exist in my source directory. Of course I can add a line to my .gitignore, but it seems a cleaner and more modern strategy to simply not write anything into the source directory during the build process.

**Comment by Cerno-b** [comment]
@wagenadl I'd like to add the use case of PyBind11 that creates a build directory for its C++ compilation and a top-level `.egg-info` directory for the wheel build. It would make for a much cleaner project setup if it was possible to generate the `.egg-info` where the other build stuff is.

**Comment by jaraco** [comment]
There's some confusion about this issue. The reported issue is about the `.eggs` directory and not the `.egg-info` directory. The former is where Setuptools will download and build/install eggs for temporary use as defined by setup_requires and tests_require if not already installed. This functionality is deprecated and unused when using Setuptools as a build backend, and there's not going to be any changes accepted to expand the scope of this behavior.

There might be a separate issue regarding the `.egg-info` directory. I agree that Setuptools shouldn't be creating this file in the source folder. And in fact, it probably doesn't need to create it at all. Probably what should happen is the backend should be re-designed to generate an sdist with the metadata in a PKG-INFO file and/or pyproject.toml and _maybe_ some other details gathered by the sdist command.

This work is probably blocked on first removing the legacy behaviors that are entangled with the backend build process (easy_install, package_index, ...). We can revisit that issue later after removing more of the tech debt.
